pandas/_libs/internals.pyx

```diff
@@ -0,0 +1,438 @@
+                self._as_slice = slc
+                self._has_slice = True
+            else:
+                arr = np.empty(0, dtype=np.int64)
+                self._as_array = arr
+                self._has_array = True
+        else:
+            # Cython memoryview interface requires ndarray to be writeable.
+            arr = np.require(val, dtype=np.int64, requirements='W')
+            assert arr.ndim == 1
+            self._as_array = arr
+            self._has_array = True
+
+    def __str__(self):
+        cdef slice s = self._ensure_has_slice()
+        if s is not None:
+            v = self._as_slice
+        else:
+            v = self._as_array
+
+        return '%s(%r)' % (self.__class__.__name__, v)
+
+    __repr__ = __str__
+
+    def __len__(self):
+        cdef slice s = self._ensure_has_slice()
+        if s is not None:
+            return slice_len(s)
+        else:
+            return len(self._as_array)
+
+    def __iter__(self):
+        cdef slice s = self._ensure_has_slice()
+        cdef Py_ssize_t start, stop, step, _
+        if s is not None:
+            start, stop, step, _ = slice_get_indices_ex(s)
+            return iter(range(start, stop, step))
+        else:
+            return iter(self._as_array)
+
+    @property
+    def as_slice(self):
+        cdef slice s = self._ensure_has_slice()
+        if s is None:
+            raise TypeError('Not slice-like')
+        else:
+            return s
+
+    @property
+    def indexer(self):
+        cdef slice s = self._ensure_has_slice()
+        if s is not None:
+            return s
+        else:
+            return self._as_array
+
+    def isin(self, arr):
+        from pandas.core.index import Int64Index
+        return Int64Index(self.as_array, copy=False).isin(arr)
+
+    @property
+    def as_array(self):
+        cdef Py_ssize_t start, stop, end, _
+        if not self._has_array:
+            start, stop, step, _ = slice_get_indices_ex(self._as_slice)
+            self._as_array = np.arange(start, stop, step,
+                                       dtype=np.int64)
+            self._has_array = True
+        return self._as_array
+
+    @property
+    def is_slice_like(self):
+        cdef slice s = self._ensure_has_slice()
+        return s is not None
+
+    def __getitem__(self, loc):
+        cdef slice s = self._ensure_has_slice()
+        if s is not None:
+            val = slice_getitem(s, loc)
+        else:
+            val = self._as_array[loc]
+
+        if not isinstance(val, slice) and val.ndim == 0:
+            return val
+
+        return BlockPlacement(val)
+
+    def delete(self, loc):
+        return BlockPlacement(np.delete(self.as_array, loc, axis=0))
+
+    def append(self, others):
+        if len(others) == 0:
+            return self
+
+        return BlockPlacement(np.concatenate([self.as_array] +
+                                             [o.as_array for o in others]))
+
+    cdef iadd(self, other):
```

Comment:
these should prob be typed

Reply:
Yep.  Bunch of small things like this I'll do in the next pass.  For now it's pure cut/paste.